gecode/int/distinct/cbs.hpp

```diff
@@ -0,0 +1,202 @@
+ *
+ * Factors precomputed for every value in the domain of x. Thoses factors are
+ * used to compute the Minc and Brégman upper bound for the permanent in
+ * counting base search
+ */
+const int MAX_MINC_FACTORS = 400;
+extern const double mincFactors[MAX_MINC_FACTORS];
+
+static double getMincFactor(int domSize) {
+  return mincFactors[domSize - 1];
+}
+
+/**
+ * \brief Liang and Bai factors
+ *
+ * Factors precomputed for every index and domain size in x. Thoses factors
+ * are used to compute the Liang and Bai upper bound for the permanent in
+ * counting base search
+ */
+const int WIDTH_LIANG_BAI_FACTORS = 400;
+extern const double liangBaiFactors[WIDTH_LIANG_BAI_FACTORS * WIDTH_LIANG_BAI_FACTORS];
+
+static double getLiangBaiFactor(int index, int domSize) {
+  return liangBaiFactors[index*WIDTH_LIANG_BAI_FACTORS + domSize - 1];
+}
+
+/**
+  * \brief Mapping of each value to its permanent update
+  *
+  */
+class ValToUpdate {
+private:
+  const size_t width;
+  const int minVal;
+  std::vector<double> mincUpdate;
+  std::vector<double> liangUpdate;
+public:
+  template<class View>
+  ValToUpdate(const ViewArray<View>& x, int minDomVal, int maxDomVal)
+    : width(maxDomVal - minDomVal + 1), minVal(minDomVal),
+      mincUpdate(width, 1.0),
+      liangUpdate(width, 1.0) {
+
+    for (int i=0; i<x.size(); i++) {
+      if (x[i].assigned()) continue;
+      size_t s = x[i].size();
+      for (ViewValues<View> val(x[i]); val(); ++val) {
+        int idx = val.val() - minVal;
+        mincUpdate[idx] *= getMincFactor(s-1) / getMincFactor(s);
+        liangUpdate[idx] *= getLiangBaiFactor(i, s-1) / getLiangBaiFactor(i, s);
+      }
+    }
+  }
+  // Gives the update we have to apply to the Minc and Brégman estimation of the
+  // permanent if we fix a variable of cardinalty "varSize" to the value "val".
+  double getMincUpdate(int val, unsigned int varSize) const {
+    return mincUpdate[val-minVal] / getMincFactor(varSize-1);
+  }
+  // Gives the update we have to apply to the Liang and Bai estimation of the
+  // permanent if we fix a variable of cardinalty "varSize" to the value "val".
+  double getLiangUpdate(int val, unsigned int idx, unsigned int varSize) const {
+    return liangUpdate[val-minVal] / getLiangBaiFactor(idx, varSize-1);
+  }
+};
+
+struct UB {
+  double minc;
+  double liangBai;
+};
+
+
+template<class View>
+void cbsdistinct(Space&, unsigned int prop_id, const ViewArray<View>& x,
+                 Propagator::SendMarginal send) {
+  // Computation of Minc and Brégman and Liang and Bai upper bounds for
+  // the permanent of the whole constraint
+  UB ub{1,1};
+  for (int i=0; i<x.size(); i++) {
+    unsigned int s = x[i].size();
+    if (s >= MAX_MINC_FACTORS || s >= WIDTH_LIANG_BAI_FACTORS)
+      throw Gecode::Exception("Int::Distinct::cbsdistinct",
+                              "Variable cardinality too big for using counting-based"
+                              "search with distinct constraints");
```

Comment:
If variable cardinalities are too big (here above 400), we throw an exception because we didn't precompute factors above that.

Reply:
Please wrap multiple conditions with || and && into parantheses.

Otherwise that's fine.